src/charm.py

```diff
@@ -0,0 +1,245 @@
+
+        # Handle basic charm lifecycle
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(
+            self.on.superset_pebble_ready, self._on_pebble_ready
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.restart_action, self._on_restart)
+
+        # Handle Ingress
+        self._require_nginx_route()
+
+    def _require_nginx_route(self):
+        """Require nginx-route relation based on current configuration."""
+        require_nginx_route(
+            charm=self,
+            service_hostname=self.external_hostname,
+            service_name=self.app.name,
+            service_port=APPLICATION_PORT,
+            tls_secret_name=self.config["tls-secret-name"],
+            backend_protocol="HTTPS",
+        )
+
+    @log_event_handler(logger)
+    def _on_install(self, event):
+        """Install application.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.unit.status = MaintenanceStatus(f"installing {APP_NAME}")
+
+    @log_event_handler(logger)
+    def _on_pebble_ready(self, event: PebbleReadyEvent):
+        """Define and start a workload using the Pebble API.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self._update(event)
+
+    @log_event_handler(logger)
+    def _on_config_changed(self, event: ConfigChangedEvent):
+        """Handle changed configuration.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.unit.status = WaitingStatus(f"configuring {APP_NAME}")
+        self._update(event)
+
+    def _restart_application(self, container):
+        """Restart application.
+
+        Args:
+            container: application container
+        """
+        self.unit.status = MaintenanceStatus(f"restarting {APP_NAME}")
+        container.restart(self.name)
+        self.unit.status = ActiveStatus()
+
+    def ready_to_start(self):
+        """Check if peer relation is established.
+
+        Returns:
+            True if peer relation established, else False.
+        """
+        if not self._state.is_ready():
+            self.unit.status = WaitingStatus("Waiting for peer relation.")
+            return False
+
+        return True
+
+    @log_event_handler(logger)
+    def _on_restart(self, event):
+        """Restart application, action handler.
+
+        Args:
+            event:The event triggered by the restart action
+        """
+        container = self.unit.get_container(self.name)
+        if not container.can_connect():
+            event.defer()
+            return
+
+        self.unit.status = MaintenanceStatus(f"restarting {APP_NAME}")
+        self._restart_application(container)
+
+        event.set_results({"result": f"{APP_NAME} successfully restarted"})
+
+    def _load_files(self, container):
```

Comment:
Does this method need to be part of the class since it is not using ```self```?

Reply:
No, I suppose there is not :) moved to utils.